common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -246,62 +272,194 @@ func (s *localStoreCertProvider) getClientTLSSettings(isWorker bool) *config.Cli
+
+	expiring = make(CertExpirationMap)
+	expired = make(CertExpirationMap)
+	when := time.Now().UTC().Add(fromNow)
+
+	checkError := checkTLSCertForExpiration(s.FetchServerCertificate, when, expiring, expired)
+	err = appendError(err, checkError)
+	checkError = checkTLSCertForExpiration(
+		func() (*tls.Certificate, error) { return s.FetchClientCertificate(false) },
+		when, expiring, expired)
+	err = appendError(err, checkError)
+	checkError = checkTLSCertForExpiration(
+		func() (*tls.Certificate, error) { return s.FetchClientCertificate(true) },
+		when, expiring, expired)
+	err = appendError(err, checkError)
+
+	// load CA certs, so that they are cached in memory
+	checkError = loadCAsAndCaptureErrors(s.FetchClientCAs)
+	err = appendError(err, checkError)
+	checkCertsForExpiration(s.clientCACerts, when, expiring, expired)
+
+	checkError = loadCAsAndCaptureErrors(func() (*x509.CertPool, error) {
+		return s.FetchServerRootCAsForClient(false)
+	})
+	err = appendError(err, checkError)
+	checkCertsForExpiration(s.serverCACerts, when, expiring, expired)
+
+	checkError = loadCAsAndCaptureErrors(func() (*x509.CertPool, error) {
+		return s.FetchServerRootCAsForClient(true)
+	})
+	err = appendError(err, checkError)
+	checkCertsForExpiration(s.serverCACertsWorker, when, expiring, expired)
+
+	return expiring, expired, err
+}
 
-	for _, ca := range caData {
-		if ca == "" {
-			continue
-		}
+func checkTLSCertForExpiration(
+	fetchCert tlsCertFetcher,
+	time time.Time,
+	expiring CertExpirationMap,
+	expired CertExpirationMap,
+) error {
 
-		caBytes, err := base64.StdEncoding.DecodeString(ca)
-		if err != nil {
-			return nil, fmt.Errorf("failed to decode ca cert: %v", err)
-		}
+	return fetchAndCheckCertForExpiration(
+		func() (*x509.Certificate, error) { return fetchAndParseTLSCert(fetchCert) },
+		time, expiring, expired)
+}
 
-		if !caPool.AppendCertsFromPEM(caBytes) {
-			return nil, errors.New("unknown failure constructing cert pool for ca")
-		}
+func fetchAndParseTLSCert(fetchCert tlsCertFetcher) (*x509.Certificate, error) {
+	cert, err := fetchCert()
+	if err != nil {
+		return nil, err
+	}
+	if cert == nil {
+		return nil, nil
+	}
+	return x509.ParseCertificate(cert.Certificate[0])
+}
 
-		atLeastOneCert = true
+func fetchAndCheckCertForExpiration(
+	fetchCert x509CertFetcher,
+	when time.Time,
+	expiring CertExpirationMap,
+	expired CertExpirationMap,
+) error {
+
+	cert, err := fetchCert()
+	checkCertForExpiration(cert, when, expiring, expired)
+	return err
+}
+
+func checkCertsForExpiration(
+	certs []*x509.Certificate,
+	time time.Time,
+	expiring CertExpirationMap,
+	expired CertExpirationMap,
+) {
+
+	for _, cert := range certs {
+		checkCertForExpiration(cert, time, expiring, expired)
 	}
+}
 
-	if !atLeastOneCert {
-		return nil, nil
+func checkCertForExpiration(
+	cert *x509.Certificate,
+	when time.Time,
+	expiring CertExpirationMap,
+	expired CertExpirationMap,
+) {
+
+	if cert != nil && cert.NotAfter.Before(when) {
```

Comment:
move to separate function called willCertExpireBeforeGivenTime(time.Time) to make it easier to read